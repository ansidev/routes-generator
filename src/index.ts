import path from 'path'
import sortRoutes from './sort-routes'

export interface Route {
  path: string
  file: string
  name: string
  children?: Route[]
}

function groupRoutes(
  routes: Set<Route>,
  rootRoutes: Set<Route> = routes
): Set<Route> {
  for (const route of routes) {
    const childRoutes: Set<Route> = new Set()
    for (const _route of routes) {
      if (_route.path.startsWith(`${route.path}/`)) {
        routes.delete(_route)
        rootRoutes.delete(_route)
        childRoutes.add(_route)
      }
    }
    if (childRoutes.size > 0) {
      route.children = sortRoutes(
        [...groupRoutes(childRoutes, rootRoutes)].map(childRoute => {
          let childPath = childRoute.path.slice(route.path.length + 1)
          if (childPath === 'index') {
            childPath = ''
          }
          return {
            ...childRoute,
            path: childPath
          }
        })
      ).map(sorted => sorted.route)
    }
  }
  return new Set(sortRoutes([...routes]).map(({ route }) => ({
    ...route,
    path: route.path.replace(/\/index$/, '')
  })))
}

export function toRoutes(
  files: Array<string> | Set<string>,
  { cwd = '' } = {}
): Route[] {
  // Sort files in ASC so that we handle shorter path first
  files = [...files].sort()
  const routes: Set<Route> = new Set()
  for (let filePath of files) {
    filePath = slash(filePath)
    const routePath = filePathToRoutePath(filePath)
    routes.add({
      path: routePath,
      file: slash(path.join(cwd, filePath)),
      name: getRouteName(filePath)
    })
  }
  const groupedRoutes = groupRoutes(routes)
  return [...groupedRoutes]
}

function filePathToRoutePath(filePath: string) {
  const routePath = `/${filePath}`
    .replace(/\.[a-z]+$/i, '') // Remove extension
    .replace(/\[([^\]]+)\]/g, ':$1')
  if (routePath === '/index') {
    return '/'
  }
  return routePath
}

function getRouteName(filePath: string) {
  return filePath.replace(/[^a-z0-9_-]/gi, '-')
}

/**
 * Convert windows back slash to forward slash
 */
function slash(filepath: string) {
  return filepath.replace(/\\/g, '/')
}
