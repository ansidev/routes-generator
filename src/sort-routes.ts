interface Route {
  path: string
}

const paramRe = /^:(.+)/

const ROOT_POINTS = 100 // `/` path goes top
const SEGMENT_POINTS = 4
const STATIC_POINTS = 3
const DYNAMIC_POINTS = 2
const SPLAT_PENALTY = 1

const isRootSegment = (segment: string) => segment === ''
const isDynamic = (segment: string) => paramRe.test(segment)
const isSplat = (segment: string) => segment === '*'

function segmentize(uri: string) {
  return (
    uri
      // strip starting/ending slashes
      .replace(/\/{2,}/, '/')
      .replace(/(^\/|\/$)/g, '')
      .split('/')
  )
}

function rankRoute<T extends Route>(route: T, index: number) {
  let score = segmentize(route.path).reduce((score, segment) => {
    score += SEGMENT_POINTS
    if (isRootSegment(segment)) score += ROOT_POINTS
    else if (isDynamic(segment)) score += DYNAMIC_POINTS
    else if (isSplat(segment)) score -= SEGMENT_POINTS + SPLAT_PENALTY
    else score += STATIC_POINTS
    return score
  }, 0)

  return {
    route,
    score,
    index
  }
}

function sortRoutes<T extends Route>(routes: T[]) {
  return routes
    .map(rankRoute)
    .sort((a, b) =>
      a.score < b.score ? 1 : a.score > b.score ? -1 : a.index - b.index
    )
}

export default sortRoutes
