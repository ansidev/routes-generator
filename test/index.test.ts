import { toRoutes } from '../src'

test('simple', () => {
  expect(toRoutes(['index.vue', 'about.vue', 'user/index.vue'])).toMatchSnapshot()
})

test('directory', () => {
  expect(
    toRoutes(['index.vue', 'foo/bar.vue', 'foo/baz.vue'])
  ).toMatchSnapshot()
})

test('nesting routes', () => {
  expect(toRoutes(['index.vue', 'foo.vue', 'foo/bar.vue', 'foo/baz.vue']))
})

test('complex example', () => {
  expect(
    toRoutes([
      '[user].vue',
      '[user]/index.vue',
      '[user]/profile.vue',
      'explore.vue',
      'explore/tweets.vue',
      'post/[post].vue',
      'post/[post]/index.vue',
      'post/[post]/[comment].vue',
      'post/[post]/[comment]/index.vue',
      'post/[post]/[comment]/likes.vue',
      'post/[post]/[comment]/replies.vue'
    ])
  ).toMatchSnapshot()
})
