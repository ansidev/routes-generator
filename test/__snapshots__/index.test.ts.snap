// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`complex example 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "component": "post/[post]/index.vue",
        "name": "post--post--index-vue",
        "path": "",
      },
      Object {
        "children": Array [
          Object {
            "component": "post/[post]/[comment]/index.vue",
            "name": "post--post---comment--index-vue",
            "path": "",
          },
          Object {
            "component": "post/[post]/[comment]/likes.vue",
            "name": "post--post---comment--likes-vue",
            "path": "likes",
          },
          Object {
            "component": "post/[post]/[comment]/replies.vue",
            "name": "post--post---comment--replies-vue",
            "path": "replies",
          },
        ],
        "component": "post/[post]/[comment].vue",
        "name": "post--post---comment--vue",
        "path": ":comment",
      },
    ],
    "component": "post/[post].vue",
    "name": "post--post--vue",
    "path": "/post/:post",
  },
  Object {
    "children": Array [
      Object {
        "component": "explore/tweets.vue",
        "name": "explore-tweets-vue",
        "path": "tweets",
      },
    ],
    "component": "explore.vue",
    "name": "explore-vue",
    "path": "/explore",
  },
  Object {
    "children": Array [
      Object {
        "component": "[user]/index.vue",
        "name": "-user--index-vue",
        "path": "",
      },
      Object {
        "component": "[user]/profile.vue",
        "name": "-user--profile-vue",
        "path": "profile",
      },
    ],
    "component": "[user].vue",
    "name": "-user--vue",
    "path": "/:user",
  },
]
`;

exports[`directory 1`] = `
Array [
  Object {
    "component": "index.vue",
    "name": "index-vue",
    "path": "/",
  },
  Object {
    "component": "foo/bar.vue",
    "name": "foo-bar-vue",
    "path": "/foo/bar",
  },
  Object {
    "component": "foo/baz.vue",
    "name": "foo-baz-vue",
    "path": "/foo/baz",
  },
]
`;

exports[`simple 1`] = `
Array [
  Object {
    "component": "index.vue",
    "name": "index-vue",
    "path": "/",
  },
  Object {
    "component": "about.vue",
    "name": "about-vue",
    "path": "/about",
  },
]
`;
